sdk/resourcemanager/azure-resourcemanager-authorization/src/main/java/com/azure/resourcemanager/authorization/implementation/ServicePrincipalImpl.java

```diff
@@ -142,20 +154,31 @@ private Mono<ServicePrincipal> submitCredentialsAsync(final ServicePrincipal sp)
 //                                new ServicePrincipalsRemoveKeyRequestBody()
 //                                    .withKeyId(UUID.fromString(id)))),
                     Flux.fromIterable(passwordCredentialsToCreate)
-                        .flatMap(passwordCredential ->
-                            manager().serviceClient().getServicePrincipals()
-                                .addPasswordAsync(id(),
-                                    new ServicePrincipalsAddPasswordRequestBodyInner()
-                                        .withPasswordCredential(passwordCredential.innerModel()))
-                                .doOnNext(passwordCredential::setInner)
-                        )
+                        .flatMap(passwordCredential -> {
+                            Mono<MicrosoftGraphPasswordCredentialInner> monoAddPassword =
+                                manager().serviceClient().getServicePrincipals()
+                                    .addPasswordAsync(id(),
+                                        new ServicePrincipalsAddPasswordRequestBodyInner()
+                                            .withPasswordCredential(passwordCredential.innerModel()));
+                            if (retry != null) {
+                                monoAddPassword = monoAddPassword.retryWhen(retry);
+                            }
+                            monoAddPassword = monoAddPassword.doOnNext(passwordCredential::setInner);
+                            return monoAddPassword;
+                        })
 //                    Flux.fromIterable(passwordCredentialsToDelete)
 //                        .flatMap(id -> manager().serviceClient().getServicePrincipals()
 //                            .removePasswordAsync(id(),
 //                                new ServicePrincipalsRemovePasswordRequestBody()
 //                                    .withKeyId(UUID.fromString(id))))
             )
-            .then(refreshAsync());
+            .then(Mono.defer(() -> {
```

Comment:
Irrelevant: maybe we can put the refresh logic outside `submitCredentialsAsync`, since it's related more closely to refresh of `ServicePrincipal` rather than submitting credentials

Reply:
From what I remember, the inner model has properties like `List<MicrosoftGraphPasswordCredentialInner> passwordCredentials`, that would be different after `submitCredentialsAsync` added some passwords; hence this refresh.

Reply:
The password could be null here and refresh would be called anyway, so I guess it's more related to refresh service principal itself? (just minor semantic opinion)

Reply:
My guess is that if no password/credential is set, the refresh can be skipped (so it could be a minor mistake here). But didn't really want to try it (in case I break something I am not aware of...)